py/desispec/io/spectra.py

```
@@ -389,16 +393,40 @@ def read_spectra(
     hdus.close()
     duration = time.time() - t0
     log.info(iotime.format("read", infile, duration))
-
+    
+    model_targetids = None # for the sanity checks in the end
+    if return_models:
+        rrmodel_file = infile.replace('coadd','rrmodel')
+        t0 = time.time()
+        if model is None:
+            model = dict()
+        if os.path.exists(rrmodel_file):
+            rrhdus = fitsio.FITS(rrmodel_file, mode="r")
+            model_targetids = Table.read(rrmodel_file, hdu="REDSHIFTS")["TARGETID"] # for sanity check
+            ind_models = []
+            # getting indices of model extensions
+            for k in range(len(rrhdus)):
+                hdu_name = fitsio.FITS(rrmodel_file)[k].get_extname()
+                if "MODEL" in hdu_name:
+                    ind_models.append(k)
+            for ind, band in zip(ind_models,bands):
+                log.debug('Reading %s_MODEL'%(band.upper()))
+                model[band] = _read_image(rrhdus, ind, np.float32, rows=rows)
+            rrhdus.close()
+            duration = time.time() - t0
+            log.info(iotime.format("read", rrmodel_file, duration))
+        else:
+            print(f'ERROR: {rrmodel_file} does not exist, so not returning redrock models')
```

I suggest that this should be a fatal error: if the user requests `return_models=True` and there are no models to return, let's raise an exception rather than just printing an error message and not returning any models.  Anything printed here should use `log.critical` instead of `print`.

Also, if `return_models` is True, consider checking for the existence of the model file earlier in the code so that it would crash before reading everything else, rather than spending severals seconds reading the spectra and then crashing at the end due to a missing model.

---

I moved the checking if file exists logic near the infile check in the beginning, and now raise the IOError, thanks for the suggestion! I think, I don't need log.critical here in this case.